### ISSUE TITLE
Java `--add-opens` should not be passed to javac; breaks with `-Werror`

### ISSUE THREAD

**Issue by timothyg-stripe** [bazelbuild/bazel#19876]
### Description of the bug:

If a target is declared with an `add_opens` attribute, it appears that the `--add-opens` flag is getting passed to `javac`. But this is incorrect, as `--add-opens` is a runtime-only flag. The consequence is that if `-Werror` is part of `javacopts` and JDK ≥17 is used, then build will fail. See reproduction case below.

### Which category does this issue belong to?

Java Rules

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

https://github.com/timothyg-stripe/bazel-add-opens-repro

```
bazel-add-opens-repro timothyg$ bazel version
Build label: 6.3.2
Build target: bazel-out/darwin_arm64-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Tue Aug 8 15:51:44 2023 (1691509904)
Build timestamp: 1691509904
Build timestamp as int: 1691509904

bazel-add-opens-repro timothyg$ cat .bazelrc
build --java_language_version=17 --javacopt=-Werror

bazel-add-opens-repro timothyg$ bazel build //:TestMain
INFO: Analyzed target //:TestMain (1 packages loaded, 3 targets configured).
INFO: Found 1 target...
ERROR: ~/bazel-add-opens-repro/BUILD.bazel:1:13: Building libtest.jar (1 source file) failed: (Exit 1): java failed: error executing command (from target //:test) external/remotejdk17_macos_aarch64/bin/java '--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED' '--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED' ... (remaining 18 arguments skipped)
warning: [options] --add-opens has no effect at compile time
error: warnings found and -Werror specified
Target //:TestMain failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.782s, Critical Path: 0.66s
INFO: 4 processes: 3 internal, 1 worker.
FAILED: Build did NOT complete successfully
```

cc @cushon who wrote 2217b13cae4110b0e2b8fe6a283a9b6dfbf150e8

### Which operating system are you running Bazel on?

macOS

### What is the output of `bazel info release`?

release 6.3.2

**Comment by cushon** [comment]
Thanks for the report, I noticed this in https://github.com/bazelbuild/bazel/issues/19850#issuecomment-1768713610 and have a pending fix

**Comment by timothyg-stripe** [comment]
Lucky timing! Could you also backport the PR to 7.x and also ideally 6.x if possible?

**Comment by cushon** [comment]
@bazel-io flag

**Comment by iancha1992** [comment]
@timothyg-stripe We can cherry-pick this into 7.0.0, but unfortunately won't be able to cherry-pick into 6.4.0 because we're scheduled to do the final release tomorrow. But if this is critical and have to be included for 6.4.0, then let us know. Thank you!

cc: @bazelbuild/triage

**Comment by iancha1992** [comment]
@bazel-io fork 7.0.0

**Comment by timothyg-stripe** [comment]
@iancha1992 that's fine, thanks for letting me know!

**Comment by timothyg-stripe** [comment]
@iancha1992 One more question: are future bugfix releases planned for 6.x? like 6.4.1 / 6.5.0? Or will the release of 7.0.0 mean that 6.x is no longer supported

**Comment by guw** [comment]
@cushon I am not sure this should be merged. I have a concern with `--add-opens` being needed by annotation processors.

**Comment by cushon** [comment]
@guw this PR should be a no-op, passing `--add-opens` as a command line argument to javac doesn't do anything. If an annotation processor was reflecting on internal JDK APIs, it would need `--add-opens` to be passed as a runtime JVM flag to javac, which is still possible.

i.e. this is the difference between `javac -J--add-opens= Foo.java` (`-J` means the flag is passed as a VM argument) and `javac --add-opens=... Foo.java` (without `-J` it's a regular javac flag).

**Comment by Wyverald** [comment]
This is fixed by https://github.com/bazelbuild/bazel/commit/367994ab9f3bb8d06a39d6bd0c139f00dcf47f3e, right?

**Comment by iancha1992** [comment]
Cherry-picked in https://github.com/bazelbuild/bazel/pull/20016

**Comment by timothyg-stripe** [comment]
@bazel-io fork 6.0.0

**Comment by iancha1992** [comment]
@timothyg-stripe this will be included in the release-7.0.0 which is set to release its RC3 next week. We will keep you updated! Thanks :)

**Comment by timothyg-stripe** [comment]
@iancha1992 Would it be possible to cherry-pick into 6.0.0 also? It will likely take us some time to upgrade to 7.0.0.

**Comment by Wyverald** [comment]
We don't have plans to create a 6.5.0 as of yet; but can consider it if more backport requests show up.

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.0.0 RC5](https://releases.bazel.build/7.0.0/rc5/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!

**Comment by keertk** [comment]
@bazel-io fork 6.5.0

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 6.5.0 RC1](https://releases.bazel.build/6.5.0/rc1/index.html) as well. Please test out the release candidate and report any issues as soon as possible. Thanks!
